{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "c6deb1bf",
   "metadata": {},
   "source": [
    "## Anna Jazayeri 131661209\n",
    "### Workshop 5\n",
    "### Text Mining BDM 550"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "b18a3507",
   "metadata": {},
   "outputs": [],
   "source": [
    "import zipfile\n",
    "from nltk.tokenize import word_tokenize\n",
    "from nltk.corpus import stopwords\n",
    "import os\n",
    "import re\n",
    "import nltk\n",
    "from sklearn.feature_extraction.text import TfidfVectorizer\n",
    "from sklearn.feature_extraction.text import CountVectorizer\n",
    "from sklearn.naive_bayes import MultinomialNB\n",
    "from sklearn.metrics import classification_report\n",
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.metrics import accuracy_score\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "35a27c58",
   "metadata": {},
   "source": [
    "import numpy as np\n",
    "import re\n",
    "\n",
    "from sklearn.feature_extraction.text import TfidfTransformer\n",
    "import pandas as pd\n",
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.naive_bayes import *\n",
    "from sklearn.metrics import classification_report, confusion_matrix, accuracy_score\n",
    "from sklearn.ensemble import RandomForestClassifier\n",
    "from sklearn.tree import DecisionTreeClassifier\n",
    "from nltk.corpus import PlaintextCorpusReader"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "24580539",
   "metadata": {},
   "source": [
    "## <span style=\"color:blue;\">Task 1: Preprocess the text to remove any stop words or punctuations.</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "5b668a74",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Accuracy: 0.37\n",
      "Classification Report:\n",
      "                   precision    recall  f1-score   support\n",
      "\n",
      "             ###       1.00      1.00      1.00         5\n",
      "            AIMX       0.00      0.00       nan         2\n",
      "         AIMX\tin       0.67      0.67      0.67         3\n",
      "        AIMX\tour       0.00       nan       nan         0\n",
      "       AIMX\tthis        nan      0.00       nan         2\n",
      "            BASE       1.00      0.50      0.67         2\n",
      "      BASE\tthese        nan      0.00       nan         1\n",
      "            CONT       0.50      0.50      0.50         2\n",
      "          CONT\ta       0.00       nan       nan         0\n",
      "    CONT\tclearly        nan      0.00       nan         1\n",
      "    CONT\thowever        nan      0.00       nan         1\n",
      "         CONT\tin        nan      0.00       nan         2\n",
      "         CONT\tso       0.00       nan       nan         0\n",
      "        CONT\tthe        nan      0.00       nan         1\n",
      "            MISC       0.46      0.48      0.47        27\n",
      "   MISC\tCITATION        nan      0.00       nan         1\n",
      "          MISC\ta       0.00       nan       nan         0\n",
      "   MISC\talthough       0.00       nan       nan         0\n",
      "         MISC\tan       0.00       nan       nan         0\n",
      " MISC\tclustering        nan      0.00       nan         2\n",
      " MISC\testimation        nan      0.00       nan         1\n",
      "    MISC\tfinding       0.00       nan       nan         0\n",
      "        MISC\tfor       1.00      0.67      0.80         3\n",
      "       MISC\tfrom        nan      0.00       nan         2\n",
      "      MISC\tgiven        nan      0.00       nan         2\n",
      "         MISC\tin        nan      0.00       nan         1\n",
      "    MISC\tmessage        nan      0.00       nan         1\n",
      "   MISC\tmoreover       0.00       nan       nan         0\n",
      "         MISC\ton        nan      0.00       nan         1\n",
      "      MISC\tsince       1.00      1.00      1.00         1\n",
      "MISC\tstatistical       0.00       nan       nan         0\n",
      "        MISC\tthe       0.50      0.50      0.50         2\n",
      "      MISC\ttheir        nan      0.00       nan         1\n",
      "      MISC\tthese       0.00       nan       nan         0\n",
      "       MISC\tthis       0.50      1.00      0.67         1\n",
      "       MISC\tthus       0.33      1.00      0.50         1\n",
      "   MISC\tunifying       0.00       nan       nan         0\n",
      "            OWNX       0.25      0.50      0.33        10\n",
      "   OWNX\talthough        nan      0.00       nan         1\n",
      "         OWNX\tas        nan      0.00       nan         1\n",
      "    OWNX\tbesides        nan      0.00       nan         1\n",
      "        OWNX\tfig        nan      0.00       nan         1\n",
      "    OWNX\tfinally       1.00      1.00      1.00         1\n",
      "      OWNX\thence        nan      0.00       nan         1\n",
      "         OWNX\tin        nan      0.00       nan         1\n",
      "         OWNX\tqa        nan      0.00       nan         1\n",
      "    OWNX\tsection        nan      0.00       nan         1\n",
      "    OWNX\tsuppose        nan      0.00       nan         1\n",
      "        OWNX\tthe        nan      0.00       nan         1\n",
      "      OWNX\tthese        nan      0.00       nan         1\n",
      "       OWNX\tthus       0.00       nan       nan         0\n",
      "      OWNX\tusing        nan      0.00       nan         1\n",
      "         OWNX\twe       0.33      0.33      0.33         9\n",
      "\n",
      "        accuracy                           0.37       101\n",
      "       macro avg       0.33      0.22      0.65       101\n",
      "    weighted avg       0.51      0.37      0.52       101\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Download and unzip the labeled and unlabeled datasets\n",
    "with zipfile.ZipFile('labeled_dataset.zip', 'r') as labeled_zip:\n",
    "    labeled_zip.extractall('labeled_dataset')\n",
    "\n",
    "with zipfile.ZipFile('unlabeled_dataset.zip', 'r') as unlabeled_zip:\n",
    "    unlabeled_zip.extractall('unlabeled_dataset')\n",
    "\n",
    "# Step 1: Load the labeled dataset\n",
    "labeled_dataset_path = 'labeled_dataset'\n",
    "unlabeled_dataset_path = 'unlabeled_dataset'\n",
    "\n",
    "def remove_stop_punc(txt):\n",
    "    # Tokenizing the text\n",
    "    tokens = nltk.word_tokenize(txt)\n",
    "    # removing the stopwords and punctuations\n",
    "    tokens = [word.lower() for word in tokens if word.isalnum() and word.lower() not in stopwords.words('english')]\n",
    "    # removing remaining punctuation using regex\n",
    "    tokens = [re.sub(r'[^\\w\\s]', '', word) for word in tokens]\n",
    "    # now the cleaned tokens must be join together as a string\n",
    "    clean_txt = ' '.join(tokens)\n",
    "    return clean_txt\n",
    "\n",
    "labeled_data = []\n",
    "labels = []\n",
    "\n",
    "for file in os.listdir(labeled_dataset_path):\n",
    "    if os.path.isfile(os.path.join(labeled_dataset_path, file)):\n",
    "        with open(os.path.join(labeled_dataset_path, file), 'r', encoding='utf-8') as f:\n",
    "            lines = f.readlines()\n",
    "            for line in lines:\n",
    "                parts = line.strip().split(' ', 1)\n",
    "                if len(parts) == 2:\n",
    "                    label, sentence = parts\n",
    "                    labeled_data.append(remove_stop_punc(sentence))\n",
    "                    labels.append(label)\n",
    "                    \n",
    "X_train, X_test, y_train, y_test = train_test_split(labeled_data, labels, test_size=0.3, random_state=42)   \n",
    "                    \n",
    "# Feature extraction using CountVectorizer\n",
    "vectorizer = CountVectorizer()\n",
    "X_train_vectorized = vectorizer.fit_transform(X_train)\n",
    "X_test_vectorized = vectorizer.transform(X_test)\n",
    "\n",
    "# Train the Naive Bayes classifier\n",
    "classifier = MultinomialNB()\n",
    "classifier.fit(X_train_vectorized, y_train)\n",
    "\n",
    "y_pred = classifier.predict(X_test_vectorized)\n",
    "accuracy = accuracy_score(y_test, y_pred)\n",
    "print(f\"Accuracy: {accuracy:.2f}\")\n",
    "print(\"Classification Report:\\n\", classification_report(y_test, y_pred, zero_division= np.nan))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "665f587f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "014ad5cc",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "###: abstract\n",
      "OWNX: transporter analyses conducted 141 organisms whose complete genome sequences available\n",
      "OWNX\twe: organism complete set membrane transport systems identified predicted functions classified protein families based transporter classification system\n",
      "OWNX\twe: organisms larger genome sizes generally possessed relatively greater number transport systems\n",
      "MISC: prokaryotes unicellular eukaryotes significant factor increase transporter content genome size greater diversity transporter types\n",
      "OWNX\twe: contrast multicellular eukaryotes greater number paralogs specific transporter families important factor increase transporter content genome size\n",
      "MISC: eukaryotic prokaryotic intracellular pathogens endosymbionts exhibited markedly limited transport capabilities\n",
      "OWNX\twe: hierarchical clustering phylogenetic profiles transporter families derived presence absence certain transporter family showed clustering patterns organisms correlated evolutionary history overall physiology lifestyles\n",
      "###: introduction\n",
      "MISC: membrane transport systems play essential roles cellular metabolism activities\n",
      "MISC: transporters function acquisition organic nutrients maintenance ion homeostasis extrusion toxic waste compounds environmental sensing cell communication important cellular functions citation\n",
      "MISC: various transport systems differ putative membrane topology energy coupling mechanisms substrate specificities citation\n",
      "###: among prevailing energy sources adenosine triphosphate phosphoenolpyruvate chemiosmotic energy form sodium ion proton electrochemical gradients\n",
      "OWNX: transporter classification system represents systematic approach classify transport systems according mode transport energy coupling mechanism molecular phylogeny substrate specificity citation citation\n",
      "OWNX: transport mode energy coupling mechanism serve primary basis classification relatively stable characteristics\n",
      "OWNX: four major classes solute transporters transporter classification system channels primary transporters secondary transporters group translocators\n",
      "MISC: transporters unknown mechanism function included distinct class\n",
      "MISC: channels transporters transport water specific types ions hydrophilic small molecules concentration electrical gradient higher rates transport lower stereospecificity transporter classes\n",
      "###: primary active transporters couple transport process primary source energy\n",
      "MISC: secondary transporters utilize ion solute electrochemical gradient motive force drive transport process\n",
      "MISC: coli lacy lactose permease citation citation probably one best characterized secondary transporters citation\n",
      "MISC: group translocators modify substrates transport process\n",
      "MISC: example coli mtla mannitol pts transporter phosphorylates exogenous mannitol using phosphoenolpyruvate phosphoryl donor energy source releases phosphate ester cell cytoplasm citation citation\n",
      "MISC: transporter class classified individual families subfamilies according function phylogeny substrate specificity citation\n",
      "MISC: since advent genomic sequencing technologies complete sequences 200 prokaryotic eukaryotic genomes published date representing wide range species archaea human\n",
      "OWNX: also additional genome sequencing projects currently underway around world citation citation\n",
      "MISC: convenient effective computational methods required handle analyze immense amount data generated sequencing projects\n",
      "OWNX: look transport proteins vital understanding metabolic capability sequenced organisms\n",
      "MISC: however often problematic annotate transport proteins current primary annotation methods occurrence large complex transporter gene families cassette superfamily citation citation major facilitator superfamily citation citation presence multiple transporter gene paralogs many organisms\n",
      "MISC: working systematic analysis cellular membrane transport systems\n",
      "MISC: previously reported comprehensive analysis transport systems 18 prokaryotic organisms citation citation yeast citation\n",
      "MISC: expand analyses 141 species compare fundamental differences membrane transport systems prokaryotes eukaryotes\n",
      "MISC: phylogenetic profiling transporter families predicted substrates utilized investigate relevance transport capabilities overall physiology prokaryotes eukaryotes\n",
      "###: abstract\n",
      "###: central problem bioinformatics gene regulation find binding sites regulatory proteins\n",
      "AIMX: one promising approaches toward identifying short fuzzy sequence patterns comparative analysis orthologous intergenic regions related species\n",
      "MISC: analysis complicated various factors\n",
      "MISC: first one needs take phylogenetic relationship species account order distinguish conservation due occurrence functional sites spurious conservation due evolutionary proximity\n",
      "MISC: second one deal complexities multiple alignments orthologous intergenic regions one consider possibility functional sites may occur outside conserved segments\n",
      "AIMX: present new motif sampling algorithm phylogibbs runs arbitrary collections multiple local sequence alignments orthologous sequences\n",
      "OWNX\twe: algorithm searches ways arbitrary number binding sites arbitrary number transcription factors assigned multiple sequence alignments\n",
      "AIMX: binding site configurations scored bayesian probabilistic model treats aligned sequences model evolution binding sites background intergenic dna\n",
      "MISC: model takes phylogenetic relationship species alignment explicitly account\n",
      "OWNX: algorithm uses simulated annealing monte carlo sampling rigorously assign posterior probabilities binding sites reports\n",
      "OWNX: tests synthetic data real data five saccharomyces species algorithm performs significantly better four algorithms including algorithms also take phylogeny account\n",
      "OWNX: results also show contrast algorithms phylogibbs make realistic estimates reliability predictions\n",
      "OWNX\twe: tests suggest running multiple alignment single gene upstream region phylogibbs average recovers 50 percent binding sites cerevisiae specificity 50 percent 33 percent binding sites specificity 85 percent\n",
      "OWNX\twe: also tested phylogibbs collections multiple alignments intergenic regions recently annotated based data contain binding sites tf\n",
      "OWNX: compared phylogibbs results previous analysis data using six algorithms\n",
      "MISC: 16 21 tfs methods failed find significant motif phylogibbs recover motif matches literature consensus\n",
      "MISC: 11 cases disagreement results compiled lists known target genes literature found running phylogibbs regulatory regions yielded binding motif matching literature consensus one cases\n",
      "MISC: interestingly literature gene lists little overlap targets annotated based data\n",
      "MISC: phylogibbs code downloaded link link\n",
      "OWNX: full set predicted sites tests yeast available link\n",
      "###: introduction\n",
      "MISC: transcription factors proteins bind manner short dna segments commonly intergenic dna upstream gene activate suppress gene transcription\n",
      "OWNX: domains recognize collections short related dna sequences\n",
      "MISC: one generally finds although unique combination bases shared binding sites although different bases occur position clear biases distribution bases occur position binding sites\n",
      "MISC: common mathematical representation motif takes variability account weight matrix citation citation w whose components w give probabilities finding base c g position binding site\n",
      "MISC: main assumption underlying mathematical representation bases occurring different positions binding site probabilistically independent\n",
      "MISC: turn follows conditions citation assumption binding energy protein dna sum pairwise contact energies individual nucleotides protein\n",
      "OWNX: several algorithms based wm representation detect ab initio binding sites common tf collection dna sequences citation citation\n",
      "OWNX: algorithms broadly fall two classes\n",
      "OWNX: one class meme citation typical representative searches space wms wm best explain observed sequences\n",
      "MISC: class gibbs sampling algorithms gibbs motif sampler citation citation typical representative instead samples space multiple alignments small sequence segments search one likely consist samples common wm\n",
      "OWNX\twe: crucial factor success ab initio methods ratio number binding sites total amount dna collection sequences\n",
      "OWNX\twe: larger number binding sites set smaller total amount dna likely ab initio methods discover binding sites among dna sequences\n",
      "MISC\tthe: order ensure reasonable chance success one thus needs provide methods collections sequences highly enriched binding sites common tf\n",
      "OWNX: one possibility use sets upstream regions genes appear microarray experiments bound common tf experiments\n",
      "OWNX: another possibility use upstream regions orthologous genes related organisms\n",
      "OWNX: assumption regulation ancestor gene thus binding sites conserved orthologs descend\n",
      "OWNX\twe: latter approach general complicated number factors\n",
      "OWNX: searching regulatory sites sequences phylogenetically related upstream regions different genes organism one may simply look short sequence motifs overrepresented among input sequences\n",
      "OWNX: set species orthologous sequences derive sufficiently diverged one may simply choose ignore phylogenetic relationship sequences treat orthologous sequences way sequences phylogenetically related\n",
      "MISC: instance approach taken mccue et al citation citation gibbs motif sampler algorithm citation citation used upstream regions bacteria\n",
      "MISC: however approach applicable datasets containing closely related species sequences exhibit significant amounts similarity simply evolutionary proximity\n",
      "AIMX: moreover amount similarity depend phylogenetic distance species clear finding conserved sequence motifs orthologous sequences closely related species much less indicative function finding sequence motifs conserved distant species\n",
      "OWNX: one general thus distinguish conservation due functional constraints conservation due evolutionary proximity correctly phylogenetic relationship sequences taken account\n",
      "OWNX: second challenge using orthologous intergenic sequences multiple species nontrivial structure multiple alignments\n",
      "BASE: one typically finds heterogeneous pattern conservation blocks different sizes covering different subsets species interspersed sequence segments show little similarity sequences species\n",
      "AIMX: technique phylogenetic footprinting restricts attention sequence segments genome interest show significant conservation species\n",
      "MISC: conserved regions multiple genes searched common motifs variety techniques\n",
      "MISC: unclear however extent regulatory sites restricted conserved segments\n",
      "MISC: instance several studies drosophila yeast citation citation shown strong correlation experimentally annotated binding sites occur whether region conserved\n",
      "OWNX: thus least yeast flies considerable information lost focusing conserved regions\n",
      "OWNX: thus decided retain entire patchwork pattern conserved sequence blocks unaligned segments\n",
      "MISC: strategy implemented gibbs sampling approach preliminary account algorithm presented citation\n",
      "OWNX: algorithm operates arbitrary collections phylogenetically related sequences orthologous intergenic regions sequences phylogenetically related upstream regions different genes organism\n",
      "AIMX: phylogenetically related groups sequences input local multiple alignments clearly similar sequence segments aligned blocks sequence segments marginal similarity left unaligned citation\n",
      "MISC: although algorithm also take global multiple alignments input believe often force phylogenetically unrelated segments aligned blocks\n",
      "MISC: may adversely affect performance algorithm\n",
      "MISC: score putative sites within blocks aligned sequences evolutionary model takes phylogenetic relationships species account putative sites unaligned segments treated independent occurrences\n",
      "###: bayesian model defines probability distribution arbitrary placements putative binding sites multiple motifs sample monte carlo markov chain\n",
      "OWNX: first use simulated annealing search globally optimal configuration binding sites\n",
      "MISC: motifs configuration tracked sampling run estimate realistic posterior probabilities binding sites algorithm reports\n",
      "OWNX: recently number algorithms developed search regulatory motifs groups phylogenetically related sequences\n",
      "OWNX: probably first algorithm proposed generalization consensus algorithm citation called phylocon citation\n",
      "MISC: phylocon operates sets genes orthologs\n",
      "OWNX: greedy algorithm first finds ungapped alignments similar sequence segments sets orthologous sequences combines alignments different upstream regions larger alignments\n",
      "OWNX: algorithm take phylogenetic information account closely related sequences treated distantly related sequences\n",
      "MISC: drawbacks algorithm assumes motif exactly one site intergenic regions assumes site conserved orthologs\n",
      "OWNX: closely related phylogibbs approach two recent algorithms citation citation generalize meme citation take phylogenetic relationships species account\n",
      "OWNX: main difference emnem phyme phyme uses evolutionary model evolution binding sites phylogibbs takes account binding sites evolve constraints set wm whereas emnem simply assumes overall slower rate evolution binding sites background sequences\n",
      "OWNX: another difference phyme like phylogibbs treats multiple alignment flexibly emnem demands global multiple alignment\n",
      "MISC: main difference phylogibbs algorithms course phylogibbs takes motif sampling approach allows us search multiple motifs parallel whereas phyme emnem use expectation maximization search one wm time\n",
      "OWNX: following sections first describe bayesian model assigns posterior probability configuration binding sites multiple motifs assigned input sequences\n",
      "OWNX: start describing model phylogenetically unrelated sequences essentially equivalent model used gibbs motif sampler citation citation describe model extended datasets contain phylogenetically related sequences\n",
      "BASE: describe move set search state space possible configurations annealing tracking strategy use identify significant groups sites\n",
      "OWNX: present examples performance algorithms synthetic real data\n",
      "OWNX: synthetic datasets consist mixtures wm samples random sequences accordance assumptions algorithms make\n",
      "OWNX: allows us compare performance algorithms idealized situation contain complexities real data\n",
      "OWNX\twe: tests also show extent binding sites recovered idealized data function quality wms number sites available number species available phylogenetic distances\n",
      "OWNX: tests real data use 200 upstream regions saccharomyces cerevisiae known binding sites collection citation compare ability different algorithms recover sites running multiple alignments orthologs upstream regions recently sequenced saccharomyces genomes citation citation\n",
      "OWNX: finally run phylogibbs collections upstream region alignments annotated citation contain binding sites common tf based data experiments extensively compare phylogibbs results annotations citation literature\n"
     ]
    }
   ],
   "source": [
    "# Use the trained classifier to predict labels for unlabeled data\n",
    "unlabeled_data = []\n",
    "\n",
    "for file in os.listdir(unlabeled_dataset_path):\n",
    "    if os.path.isfile(os.path.join(unlabeled_dataset_path, file)):\n",
    "        with open(os.path.join(unlabeled_dataset_path, file), 'r', encoding='utf-8') as f:\n",
    "            unlabeled_data.extend(remove_stop_punc(line.strip()) for line in f.readlines())\n",
    "\n",
    "# Vectorize the unlabeled data\n",
    "unlabeled_data_vectorized = vectorizer.transform(unlabeled_data)\n",
    "\n",
    "# Predict labels for the unlabeled data\n",
    "predicted_labels = classifier.predict(unlabeled_data_vectorized)\n",
    "\n",
    "# Print or save the predicted labels for further use\n",
    "for sentence, label in zip(unlabeled_data, predicted_labels):\n",
    "    print(f\"{label}: {sentence}\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d7bb027b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "59fc48f2",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "62ecc11d",
   "metadata": {},
   "source": [
    "## <span style=\"color:blue;\">Task 2: Use TF-IDF to vectorize the sentences.</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4d0e77c1",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "951702a7",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b9e133d2",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "19599ee4",
   "metadata": {},
   "source": [
    "## <span style=\"color:blue;\">Task 3: Use Scikit learn to create a (naïve bayes) classifier.</span>\n",
    "\n",
    "<span style=\"color:blue;\">**First create a classifier (naïve bayes) to classify the given dataset into 6 categories: (AIMX, OWNX, CONT, BASE, NUMBER, and MISC).<br>Then, use the classifier to label the sentences in the unlabeled dataset.**</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3b03e17a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "90d7891a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f383fa01",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "d52ebb92",
   "metadata": {},
   "source": [
    "## <span style=\"color:blue;\">Task 4: Summarize your work and your findings in a few sentences.</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bcb2fc4e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "90f139d4",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "8e8ecbb8",
   "metadata": {},
   "source": [
    "### <span style=\"color:blue;\">BONUS [Extra 5% added to any workshops]: Use other classification approaches to label the unlabeled sentences. Evaluate your work using precision, recall, and f1 score.</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "982e055a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ada2187d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fcdfeaa7",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
